fabric-networking-handshake-v1/src/main/java/net/fabricmc/fabric/impl/networking/handshake/HandshakeMod.java

```
@@ -0,0 +1,209 @@
+import net.fabricmc.fabric.api.networking.v1.sender.PacketByteBufs;
+import net.fabricmc.fabric.api.networking.v1.sender.PacketSender;
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
+import net.fabricmc.loader.api.Version;
+import net.fabricmc.loader.api.metadata.CustomValue;
+import net.fabricmc.loader.api.metadata.CustomValue.CvType;
+import net.fabricmc.loader.api.metadata.ModMetadata;
+import net.fabricmc.loader.util.version.VersionParsingException;
+import net.fabricmc.loader.util.version.VersionPredicateParser;
+
+/**
+ * The entrypoint for the handshake mod.
+ */
+public final class HandshakeMod implements ModInitializer {
+	static final String CUSTOM_KEY = "fabric-networking-handshake";
+	static final String ID = "fabric-networking-handshake-v1";
+	static final Identifier HELLO_CHANNEL = id("hello");
+	private static final Logger LOGGER = LogManager.getLogger();
+	// todo future: add a list of legacy handlers which warns when legacy is detected
+	private Multimap<String, ModVersionReporter> remoteVersionRequirements;
+
+	/**
+	 * Creates an id with the mod's id as its namespace.
+	 *
+	 * @param name the name
+	 * @return the created id
+	 */
+	public static Identifier id(String name) {
+		return new Identifier(ID, name);
+	}
+
+	@Override
+	public void onInitialize() {
+		FabricLoader loader = FabricLoader.getInstance();
+		Multimap<String, ModVersionReporter> remoteVersionRequirements = HashMultimap.create();
+
+		for (ModContainer mod : loader.getAllMods()) {
+			ModMetadata metadata = mod.getMetadata();
+			CustomValue value = metadata.getCustomValue(CUSTOM_KEY);
+
+			if (value != null && value.getType() == CvType.OBJECT) {
+				CustomValue.CvObject cvObject = value.getAsObject();
+
+				CustomValue versionRange = cvObject.get("versionRange");
+
+				if (versionRange != null && versionRange.getType() == CvType.STRING) {
+					String st = versionRange.getAsString();
```

If I understand this correctly, it looks like the version range will be parsed only when the ModVersionReporter is called. This means that the version range has to be parsed each time, and an invalid version range string would result in all clients failing to connect. Ideally, it should be parsed once and immediately so that the server can fail fast for an invalid version range. It might require some changes in loader.

---

The parsing of version range as a predicate of semver or arbitrary version is determined by both the range string and input string, which is handled internally by the loader. As a result, I keep only the version range string.

---

Handshaking has been postponed.